Missing lock around rng access in rand.go

### Description

A rand.Rand instance is used without locking in rand.go, leading to potential panics.

### Environment

- OS: macOS
- Architecture: arm64
- Go Version: 1.23
- opentelemetry-go version: 0.29.0

### Steps To Reproduce

1. Call `Int64Histogram.Record` from multiple goroutines.
2. See error below.

```
panic: runtime error: index out of range [-1]
2024-09-12 13:11:53.296	

2024-09-12 13:11:53.296	
goroutine 1375207 [running]:
2024-09-12 13:11:53.296	
math/rand.(*rngSource).Uint64(...)
2024-09-12 13:11:53.296	
	math/rand/rng.go:249
2024-09-12 13:11:53.296	
math/rand.(*rngSource).Int63(0x2911fc8?)
2024-09-12 13:11:53.296	
	math/rand/rng.go:234 +0x85
2024-09-12 13:11:53.296	
math/rand.(*Rand).Int63(...)
2024-09-12 13:11:53.296	
	math/rand/rand.go:96
2024-09-12 13:11:53.296	
math/rand.(*Rand).Int63n(0x8c14e4ce7d6c0496?, 0x1?)
2024-09-12 13:11:53.296	
	math/rand/rand.go:125 +0x76
2024-09-12 13:11:53.296	
go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).Offer(0xc000822780, {0x2947a18, 0xc1c9c0d770}, {0xc1c9cfc160?, 0xc1b579d528?, 0x3dcfce0?}, {0x28?, 0x40810d?}, {0x0, 0x0, ...})
2024-09-12 13:11:53.296	
	go.opentelemetry.io/otel/sdk/metric@v1.29.0/internal/exemplar/rand.go:126 +0x96
2024-09-12 13:11:53.296	
go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*filteredReservoir[...]).Offer(0x291e2e0, {0x2947a18, 0xc1c9c0d770}, 0x6f0, {0x0, 0x0, 0x0})
2024-09-12 13:11:53.296	
	go.opentelemetry.io/otel/sdk/metric@v1.29.0/internal/exemplar/filtered_reservoir.go:45 +0x127
2024-09-12 13:11:53.296	
go.opentelemetry.io/otel/sdk/metric/internal/aggregate.(*valueMap[...]).measure(0x2930680, {0x2947a18, 0xc1c9c0d770}, 0x6f0, {{{0x202a300, 0xc1c9cf2d80?}}}, {0x0, 0x0, 0x0})
2024-09-12 13:11:53.296	
	go.opentelemetry.io/otel/sdk/metric@v1.29.0/internal/aggregate/sum.go:50 +0x31f
2024-09-12 13:11:53.296	
go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[...].filter.func9(0x0?, {{{0x202a300?, 0xc1c9cf2d80?}}})
2024-09-12 13:11:53.296	
	go.opentelemetry.io/otel/sdk/metric@v1.29.0/internal/aggregate/aggregate.go:72 +0x35
2024-09-12 13:11:53.296	
go.opentelemetry.io/otel/sdk/metric.(*int64Inst).aggregate(...)
2024-09-12 13:11:53.296	
	go.opentelemetry.io/otel/sdk/metric@v1.29.0/instrument.go:201
2024-09-12 13:11:53.296	
go.opentelemetry.io/otel/sdk/metric.(*int64Inst).Add(0xc000f615c0, {0x2947a18, 0xc1c9c0d770}, 0x6f0, {0xc1c9cfc180?, 0x4155ab?, 0x7a5a8d6b5c60?})
2024-09-12 13:11:53.296	
	go.opentelemetry.io/otel/sdk/metric@v1.29.0/instrument.go:191 +0x89
```

### Expected behavior

No panics from code with concurrent calls to a histogram.

Reproduction:

```go
package metric // import "go.opentelemetry.io/otel/sdk/metric"

import (
	"context"
	"sync"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"go.opentelemetry.io/otel/sdk/metric/metricdata"
)

func TestRandomExemplarConcurrentSafe(t *testing.T) {
	t.Setenv("OTEL_METRICS_EXEMPLAR_FILTER", "always_on")

	r := NewManualReader()
	p := NewMeterProvider(WithReader(r))
	m := p.Meter("exemplar-concurrency")
	// Use two instruments to get concurrent access to any shared globals.
	i0, err := m.Int64Counter("counter.0")
	require.NoError(t, err)
	i1, err := m.Int64Counter("counter.1")
	require.NoError(t, err)

	ctx := context.Background()

	const goRoutines = 10

	var wg sync.WaitGroup
	wg.Add(goRoutines)
	done := make(chan struct{})
	go func() {
		wg.Wait()
		close(done)
	}()

	go func() {
		select {
		case <-done:
			return
		default:
		}
		var out metricdata.ResourceMetrics
		require.NotPanics(t, func() {
			require.NoError(t, r.Collect(ctx, &out))
		})
	}()

	for n := 0; n < goRoutines; n++ {
		go func() {
			defer wg.Done()
			require.NotPanics(t, func() {
				i0.Add(ctx, 1)
				i1.Add(ctx, 2)
			})
		}()
	}

	<-done

	var out metricdata.ResourceMetrics
	require.NoError(t, r.Collect(ctx, &out))
	p.Shutdown(ctx)
}
```

```terminal
go test -race -count=100 -failfast -run="TestRandomExemplarConcurrentSafe" -v
=== RUN   TestRandomExemplarConcurrentSafe
--- PASS: TestRandomExemplarConcurrentSafe (0.00s)
=== RUN   TestRandomExemplarConcurrentSafe
==================
WARNING: DATA RACE
Read at 0x00c000199500 by goroutine 31:
  math/rand.(*rngSource).Uint64()
      /usr/lib/go/src/math/rand/rng.go:239 +0x30
  math/rand.(*rngSource).Int63()
      /usr/lib/go/src/math/rand/rng.go:234 +0x1cd
  math/rand.(*Rand).Int63()
      /usr/lib/go/src/math/rand/rand.go:96 +0xa7
  math/rand.(*Rand).Int63n()
      /usr/lib/go/src/math/rand/rand.go:125 +0xad
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).Offer()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:126 +0x3ac
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*filteredReservoir[go.shape.int64]).Offer()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/filtered_reservoir.go:45 +0x282
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*filteredReservoir[int64]).Offer()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/filtered_reservoir.go:42 +0x84
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.(*valueMap[go.shape.int64]).measure()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/sum.go:50 +0x4eb
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:119 +0xdb
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.Builder[go.shape.int64].filter.func9()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:72 +0x79
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).aggregate()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:201 +0xe2
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).Add()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:191 +0x76
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3.1()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:56 +0x7b
  github.com/stretchr/testify/assert.didPanic()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1201 +0xdc
  github.com/stretchr/testify/assert.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1272 +0xa4
  github.com/stretchr/testify/require.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/require/require.go:1605 +0xb3
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:55 +0x205

Previous write at 0x00c000199500 by goroutine 32:
  math/rand.(*rngSource).Uint64()
      /usr/lib/go/src/math/rand/rng.go:239 +0x44
  math/rand.(*rngSource).Int63()
      /usr/lib/go/src/math/rand/rng.go:234 +0x1cd
  math/rand.(*Rand).Int63()
      /usr/lib/go/src/math/rand/rand.go:96 +0xd8
  math/rand.(*Rand).Float64()
      /usr/lib/go/src/math/rand/rand.go:207 +0xd9
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.random()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:49 +0x7f
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).advance()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:181 +0xfc
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).reset()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:152 +0x144
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.FixedSize()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:62 +0x124
  go.opentelemetry.io/otel/sdk/metric.reservoirFunc[go.shape.int64].func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar.go:75 +0x125
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.(*valueMap[go.shape.int64]).measure()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/sum.go:45 +0x46d
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:119 +0xdb
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.Builder[go.shape.int64].filter.func9()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:72 +0x79
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).aggregate()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:201 +0xe2
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).Add()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:191 +0x76
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3.1()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:57 +0xa5
  github.com/stretchr/testify/assert.didPanic()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1201 +0xdc
  github.com/stretchr/testify/assert.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1272 +0xa4
  github.com/stretchr/testify/require.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/require/require.go:1605 +0xb3
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:55 +0x205

Goroutine 31 (running) created at:
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:53 +0x45b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44

Goroutine 32 (finished) created at:
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:53 +0x45b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44
==================
==================
WARNING: DATA RACE
Read at 0x00c000199508 by goroutine 31:
  math/rand.(*rngSource).Uint64()
      /usr/lib/go/src/math/rand/rng.go:244 +0xa4
  math/rand.(*rngSource).Int63()
      /usr/lib/go/src/math/rand/rng.go:234 +0x1cd
  math/rand.(*Rand).Int63()
      /usr/lib/go/src/math/rand/rand.go:96 +0xa7
  math/rand.(*Rand).Int63n()
      /usr/lib/go/src/math/rand/rand.go:125 +0xad
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).Offer()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:126 +0x3ac
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*filteredReservoir[go.shape.int64]).Offer()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/filtered_reservoir.go:45 +0x282
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*filteredReservoir[int64]).Offer()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/filtered_reservoir.go:42 +0x84
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.(*valueMap[go.shape.int64]).measure()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/sum.go:50 +0x4eb
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:119 +0xdb
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.Builder[go.shape.int64].filter.func9()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:72 +0x79
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).aggregate()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:201 +0xe2
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).Add()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:191 +0x76
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3.1()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:56 +0x7b
  github.com/stretchr/testify/assert.didPanic()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1201 +0xdc
  github.com/stretchr/testify/assert.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1272 +0xa4
  github.com/stretchr/testify/require.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/require/require.go:1605 +0xb3
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:55 +0x205

Previous write at 0x00c000199508 by goroutine 32:
  math/rand.(*rngSource).Uint64()
      /usr/lib/go/src/math/rand/rng.go:244 +0xbc
  math/rand.(*rngSource).Int63()
      /usr/lib/go/src/math/rand/rng.go:234 +0x1cd
  math/rand.(*Rand).Int63()
      /usr/lib/go/src/math/rand/rand.go:96 +0xd8
  math/rand.(*Rand).Float64()
      /usr/lib/go/src/math/rand/rand.go:207 +0xd9
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.random()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:49 +0x7f
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).advance()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:181 +0xfc
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.(*randRes).reset()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:152 +0x144
  go.opentelemetry.io/otel/sdk/metric/internal/exemplar.FixedSize()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/exemplar/rand.go:62 +0x124
  go.opentelemetry.io/otel/sdk/metric.reservoirFunc[go.shape.int64].func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar.go:75 +0x125
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.(*valueMap[go.shape.int64]).measure()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/sum.go:45 +0x46d
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:119 +0xdb
  go.opentelemetry.io/otel/sdk/metric/internal/aggregate.Builder[go.shape.int64].Sum.Builder[go.shape.int64].filter.func9()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/internal/aggregate/aggregate.go:72 +0x79
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).aggregate()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:201 +0xe2
  go.opentelemetry.io/otel/sdk/metric.(*int64Inst).Add()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/instrument.go:191 +0x76
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3.1()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:57 +0xa5
  github.com/stretchr/testify/assert.didPanic()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1201 +0xdc
  github.com/stretchr/testify/assert.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1272 +0xa4
  github.com/stretchr/testify/require.NotPanics()
      /home/tyler/go/pkg/mod/github.com/stretchr/testify@v1.9.0/require/require.go:1605 +0xb3
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe.func3()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:55 +0x205

Goroutine 31 (running) created at:
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:53 +0x45b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44

Goroutine 32 (finished) created at:
  go.opentelemetry.io/otel/sdk/metric.TestRandomExemplarConcurrentSafe()
      /home/tyler/go/src/go.opentelemetry.io/otel/sdk/metric/exemplar_test.go:53 +0x45b
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1743 +0x44
==================
    testing.go:1399: race detected during execution of test
--- FAIL: TestRandomExemplarConcurrentSafe (0.00s)
FAIL
exit status 1
FAIL	go.opentelemetry.io/otel/sdk/metric	0.017s
```

The key here is the duplicate instruments (i.e. `i0` and `i1`) that share access to the global. This does not fail with only one instrument.

This does not fail reliably with only one test run. Need to build multiple collections into this test to cause it to do so.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - `Logger.Enabled` in `go.opentelemetry.io/otel/log` now accepts a newly introduced `EnabledParameters` type instead of `Record`. (#5791)
 - `FilterProcessor.Enabled` in `go.opentelemetry.io/otel/sdk/log/internal/x` now accepts `EnabledParameters` instead of `Record`. (#5791)
 
+### Fixed
+
+- The race condition for multiple `FixedSize` exemplar reservoirs identified in #5814 is resolved. (#5819)
+
 <!-- Released section -->
 <!-- Don't change this section unless doing release -->
 

--- a/sdk/metric/exemplar_test.go
+++ b/sdk/metric/exemplar_test.go
@@ -10,6 +10,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
+
 	"go.opentelemetry.io/otel/sdk/metric/metricdata"
 )
 

--- a/sdk/metric/exemplar_test.go
+++ b/sdk/metric/exemplar_test.go
@@ -1,0 +1,62 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metric // import "go.opentelemetry.io/otel/sdk/metric"
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+func TestFixedSizeExemplarConcurrentSafe(t *testing.T) {
+	// Tests https://github.com/open-telemetry/opentelemetry-go/issues/5814
+
+	t.Setenv("OTEL_METRICS_EXEMPLAR_FILTER", "always_on")
+
+	r := NewManualReader()
+	m := NewMeterProvider(WithReader(r)).Meter("exemplar-concurrency")
+	// Use two instruments to get concurrent access to any shared globals.
+	i0, err := m.Int64Counter("counter.0")
+	require.NoError(t, err)
+	i1, err := m.Int64Counter("counter.1")
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	add := func() {
+		i0.Add(ctx, 1)
+		i1.Add(ctx, 2)
+	}
+
+	goRoutines := max(10, runtime.NumCPU())
+
+	var wg sync.WaitGroup
+	for n := 0; n < goRoutines; n++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					require.NotPanics(t, add)
+				}
+			}
+		}()
+	}
+
+	const collections = 100
+	var rm metricdata.ResourceMetrics
+	for c := 0; c < collections; c++ {
+		require.NotPanics(t, func() { _ = r.Collect(ctx, &rm) })
+	}
+
+	cancel()
+	wg.Wait()
+}

--- a/sdk/metric/internal/exemplar/rand.go
+++ b/sdk/metric/internal/exemplar/rand.go
@@ -7,25 +7,50 @@ import (
 	"context"
 	"math"
 	"math/rand"
-	"sync"
 	"time"
 
 	"go.opentelemetry.io/otel/attribute"
 )
 
-var (
+// FixedSize returns a [Reservoir] that samples at most k exemplars. If there
+// are k or less measurements made, the Reservoir will sample each one. If
+// there are more than k, the Reservoir will then randomly sample all
+// additional measurement with a decreasing probability.
+func FixedSize(k int) Reservoir {
+	return newRandRes(newStorage(k))
+}
+
+type randRes struct {
+	*storage
+
+	// count is the number of measurement seen.
+	count int64
+	// next is the next count that will store a measurement at a random index
+	// once the reservoir has been filled.
+	next int64
+	// w is the largest random number in a distribution that is used to compute
+	// the next next.
+	w float64
+
 	// rng is used to make sampling decisions.
 	//
 	// Do not use crypto/rand. There is no reason for the decrease in performance
 	// given this is not a security sensitive decision.
-	rng = rand.New(rand.NewSource(time.Now().UnixNano()))
-	// Ensure concurrent safe access to rng and its underlying source.
-	rngMu sync.Mutex
-)
+	rng *rand.Rand
+}
 
-// random returns, as a float64, a uniform pseudo-random number in the open
-// interval (0.0,1.0).
-func random() float64 {
+func newRandRes(s *storage) *randRes {
+	r := &randRes{
+		storage: s,
+		rng:     rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	r.reset()
+	return r
+}
+
+// randomFloat64 returns, as a float64, a uniform pseudo-random number in the
+// open interval (0.0,1.0).
+func (r *randRes) randomFloat64() float64 {
 	// TODO: This does not return a uniform number. rng.Float64 returns a
 	// uniformly random int in [0,2^53) that is divided by 2^53. Meaning it
 	// returns multiples of 2^-53, and not all floating point numbers between 0
@@ -43,37 +68,11 @@ func random() float64 {
 	//
 	// There are likely many other methods to explore here as well.
 
-	rngMu.Lock()
-	defer rngMu.Unlock()
-
-	f := rng.Float64()
+	f := r.rng.Float64()
 	for f == 0 {
-		f = rng.Float64()
+		f = r.rng.Float64()
 	}
 	return f
-}
-
-// FixedSize returns a [Reservoir] that samples at most k exemplars. If there
-// are k or less measurements made, the Reservoir will sample each one. If
-// there are more than k, the Reservoir will then randomly sample all
-// additional measurement with a decreasing probability.
-func FixedSize(k int) Reservoir {
-	r := &randRes{storage: newStorage(k)}
-	r.reset()
-	return r
-}
-
-type randRes struct {
-	*storage
-
-	// count is the number of measurement seen.
-	count int64
-	// next is the next count that will store a measurement at a random index
-	// once the reservoir has been filled.
-	next int64
-	// w is the largest random number in a distribution that is used to compute
-	// the next next.
-	w float64
 }
 
 func (r *randRes) Offer(ctx context.Context, t time.Time, n Value, a []attribute.KeyValue) {
@@ -123,7 +122,7 @@ func (r *randRes) Offer(ctx context.Context, t time.Time, n Value, a []attribute
 	} else {
 		if r.count == r.next {
 			// Overwrite a random existing measurement with the one offered.
-			idx := int(rng.Int63n(int64(cap(r.store))))
+			idx := int(r.rng.Int63n(int64(cap(r.store))))
 			r.store[idx] = newMeasurement(ctx, t, n, a)
 			r.advance()
 		}
@@ -147,7 +146,7 @@ func (r *randRes) reset() {
 	// This maps the uniform random number in (0,1) to a geometric distribution
 	// over the same interval. The mean of the distribution is inversely
 	// proportional to the storage capacity.
-	r.w = math.Exp(math.Log(random()) / float64(cap(r.store)))
+	r.w = math.Exp(math.Log(r.randomFloat64()) / float64(cap(r.store)))
 
 	r.advance()
 }
@@ -167,7 +166,7 @@ func (r *randRes) advance() {
 	// therefore the next r.w will be based on the same distribution (i.e.
 	// `max(u_1,u_2,...,u_k)`). Therefore, we can sample the next r.w by
 	// computing the next random number `u` and take r.w as `w * u^(1/k)`.
-	r.w *= math.Exp(math.Log(random()) / float64(cap(r.store)))
+	r.w *= math.Exp(math.Log(r.randomFloat64()) / float64(cap(r.store)))
 	// Use the new random number in the series to calculate the count of the
 	// next measurement that will be stored.
 	//
@@ -178,7 +177,7 @@ func (r *randRes) advance() {
 	//
 	// Important to note, the new r.next will always be at least 1 more than
 	// the last r.next.
-	r.next += int64(math.Log(random())/math.Log(1-r.w)) + 1
+	r.next += int64(math.Log(r.randomFloat64())/math.Log(1-r.w)) + 1
 }
 
 func (r *randRes) Collect(dest *[]Exemplar) {

--- a/sdk/metric/internal/exemplar/rand_test.go
+++ b/sdk/metric/internal/exemplar/rand_test.go
@@ -6,9 +6,10 @@ package exemplar
 import (
 	"context"
 	"math"
+	"math/rand"
 	"slices"
-	"sync"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 )
@@ -27,10 +28,12 @@ func TestFixedSizeSamplingCorrectness(t *testing.T) {
 	intensity := 0.1
 	sampleSize := 1000
 
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+
 	data := make([]float64, sampleSize*1000)
 	for i := range data {
 		// Generate exponentially distributed data.
-		data[i] = (-1.0 / intensity) * math.Log(random())
+		data[i] = (-1.0 / intensity) * math.Log(rng.Float64())
 	}
 	// Sort to test position bias.
 	slices.Sort(data)
@@ -49,19 +52,4 @@ func TestFixedSizeSamplingCorrectness(t *testing.T) {
 	// Check the intensity/rate of the sampled distribution is preserved
 	// ensuring no bias in our random sampling algorithm.
 	assert.InDelta(t, 1/mean, intensity, 0.02) // Within 5σ.
-}
-
-func TestRandomConcurrentSafe(t *testing.T) {
-	const goRoutines = 10
-
-	var wg sync.WaitGroup
-	for n := 0; n < goRoutines; n++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			_ = random()
-		}()
-	}
-
-	wg.Wait()
 }

